File: Day04/py/main.py
def calculate_points(cards):
    total_points = 0

    for card in cards:
        # split the card into winning numbers and your numbers
        sides = card.split("|")
        winning_numbers = sides[0].split()
        your_numbers = sides[1].split()

        points = 0
        for your_number in your_numbers:
            # if a number matches, double the points (or set to 1 if this is the first match)
            if your_number in winning_numbers:
                points = 1 if points == 0 else points * 2

        total_points += points

    return total_points

def part1():
    # read the cards from the file
    with open('input.txt', 'r') as file:
        cards = file.read().splitlines()

    # calculate the total points
    result = calculate_points(cards)
    return f'part one: {result}'

def calculate_matches(cards):
    matches = [0] * len(cards)

    for i, card in enumerate(cards):
        # split the card into winning numbers and your numbers
        sides = card.split("|")
        winning_numbers = sides[0].split()
        your_numbers = sides[1].split()

        for your_number in your_numbers:
            # if a number matches, increment the match count for this card
            if your_number in winning_numbers:
                matches[i] += 1

    return matches

def part2():
    # read the cards from the file
    with open('input.txt', 'r') as file:
        cards = file.read().splitlines()

    # calculate the number of matches for each card
    matches = calculate_matches(cards)
    cards_won = [1] * len(cards)

    for i, match in enumerate(matches):
        # for each match, add the number of cards won to the next cards
        for j in range(1, match + 1):
            if i + j < len(cards):
                cards_won[i + j] += cards_won[i]

    # calculate the total number of cards won
    total = sum(cards_won)

    return f'Part two: {total}'

def main():
    # print the results of part 1 and part 2
    print(part1())
    print(part2())

if __name__ == "__main__":
    main()

File: Day04/go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

// calculatePoints calculates the points for each card
func calculatePoints(cards []string) int {
	totalPoints := 0

	for _, card := range cards {
		// split the card into winning numbers and your numbers
		sides := strings.Split(card, "|")
		winningNumbers := strings.Fields(sides[0])
		yourNumbers := strings.Fields(sides[1])

		points := 0
		for _, yourNumber := range yourNumbers {
			for _, winningNumber := range winningNumbers {
				if yourNumber == winningNumber {
					// if a number matches, double the points (or set to 1 if this is the first match)
					if points == 0 {
						points = 1
					} else {
						points *= 2
					}
					break
				}
			}
		}
		totalPoints += points
	}

	return totalPoints
}

// part1 reads the cards from the file and calculates the total points
func part1() string {
	file, _ := os.Open("input.txt")
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var cards []string
	for scanner.Scan() {
		cards = append(cards, scanner.Text())
	}

	result := calculatePoints(cards)
	return fmt.Sprintf("part one: %d", result)
}

// calculateMatches calculates the number of matches for each card
func calculateMatches(cards []string) []int {
	matches := make([]int, len(cards))

	for i, card := range cards {
		// split the card into winning numbers and your numbers
		sides := strings.Split(card, "|")
		winningNumbers := strings.Fields(sides[0])
		yourNumbers := strings.Fields(sides[1])

		for _, yourNumber := range yourNumbers {
			for _, winningNumber := range winningNumbers {
				if yourNumber == winningNumber {
					// if a number matches, increment the match count for this card
					matches[i]++
					break
				}
			}
		}
	}

	return matches
}

// part2 reads the cards from the file and calculates the total number of cards won
func part2() string {
	file, _ := os.Open("input.txt")
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var cards []string
	for scanner.Scan() {
		cards = append(cards, scanner.Text())
	}

	matches := calculateMatches(cards)
	cardsWon := make([]int, len(cards))
	for i := range cardsWon {
		cardsWon[i] = 1
	}

	for i, match := range matches {
		// for each match, add the number of cards won to the next cards
		for j := 1; j <= match && i+j < len(cards); j++ {
			cardsWon[i+j] += cardsWon[i]
		}
	}

	total := 0
	for _, won := range cardsWon {
		total += won
	}

	return fmt.Sprintf("Part two: %d", total)
}

// main function prints the results of part 1 and part 2
func main() {
	fmt.Println(part1())
	fmt.Println(part2())
}
